src/baseAtomWithQuery.ts

```diff
@@ -63,92 +50,221 @@ export function baseAtomWithQuery<
     defaultedOptions._optimisticResults = 'optimistic'
 
     if (cachedObserver) {
+      // This is equivalent to this effect
+      // https://github.com/TanStack/query/blob/main/packages/react-query/src/useBaseQuery.ts#L94
+      // but notice that tanstack/react-query is always returning either observer.getOptimisticResult or observer.fetchOptimistic,
+      // it never uses value from observer.subscribe (and this implementation does).
+      //
+      // Yet, I think that it works as expected,
+      // because if theres an error, then resultAtom is cleared, therefore after remounting, optimisticResultAtom will be called.
       cachedObserver.setOptions(defaultedOptions, {
         listeners: false,
       })
     }
 
     return ensureStaleTime(defaultedOptions)
   })
-  if (process.env.NODE_ENV !== 'production') {
-    defaultedOptionsAtom.debugPrivate = true
-  }
-
   const observerAtom = atom((get) => {
     const client = get(clientAtom)
-    const defaultedOptions = get(defaultedOptionsAtom)
-
     const observerCache = get(observerCacheAtom)
-
     const cachedObserver = observerCache.get(client)
 
-    if (cachedObserver) return cachedObserver
-
-    const newObserver = new Observer(client, defaultedOptions)
+    if (cachedObserver) {
+      return cachedObserver
+    }
+    // Prevent recalculating this atom by getting defaultedOptions only when creating observer for the first time,
+    // later observer options will get updated via defaultedOptionsAtom.
```

Comment:
@kalijonn this is new comparing to my initial suggestion.

Reply:
this looks good.